### PR TITLE
RequestNodeMatch predicate

### PR DESCRIPTION
**Pull request by jyotimahapatra**
This PR adds the [RequestNodeMatch](https://github.com/envoyproxy/xds-relay/blob/master/api/protos/aggregation/v1/aggregation.proto#L57-L74) predicate

### CODE REVIEW

**Comment by jyotimahapatra** [comment]
PTAL @LisaLudique @eapolinario @jessicayuen

**Comment by jessicayuen** [comment]
@jyotimahapatra can you merge in #24 ?

**Comment by jyotimahapatra** [comment]
> @jyotimahapatra can you merge in #24 ?

done

File: internal/app/mapper/mapper.go [hunk 1]
```diff
@@ -63,8 +68,37 @@ func (mapper *mapper) GetKey(request v2.DiscoveryRequest) (string, error) {
...
+func isMatch(matchPredicate *matchPredicate, typeURL string, node *core.Node) (bool, error) {
+	result, err := isNodeMatch(matchPredicate, node)
+	if err != nil {
+		return false, err
+	}
+	if result {
+		return result, nil
+	}
+
+	return isRequestTypeMatch(matchPredicate, typeURL) || isAnyMatch(matchPredicate), nil
```
Comment (jessicayuen) [inline internal/app/mapper/mapper.go]:
```suggestion
	isNodeMatch, err := isNodeMatch(matchPredicate, node)
	if err != nil {
		return false, err
	}

	return isNodeMatch || isRequestTypeMatch(matchPredicate, typeURL) || isAnyMatch(matchPredicate), nil
```

File: internal/app/mapper/mapper.go [hunk 2]
```diff
@@ -89,3 +123,21 @@ func getResult(fragmentRule *rule) string {
 	stringFragment := fragmentRule.GetResult().GetStringFragment()
 	return stringFragment
 }
+
+func compare(requestNodeMatch *aggregationv1.MatchPredicate_RequestNodeMatch, nodeValue string) (bool, error) {
+	exactMatch := requestNodeMatch.GetExactMatch()
+	if exactMatch != "" {
```
Comment (jessicayuen) [inline internal/app/mapper/mapper.go]: We would get a false negative if the nodeValue is empty right? Perhaps that is desired, but something to consider. Regardless would add a test case for this scenario.

File: internal/app/mapper/mapper_test.go [hunk 3]
```diff
@@ -65,6 +70,96 @@ var postivetests = []TableEntry{
...
+			getRequestNodeRegexMatch(nodeIDField, "n....d"),
+			getResultStringFragment(),
+			clusterTypeURL,
+			stringFragment,
+		},
+	},
+	{
+		Description: "RequestNodeMatch with node cluster regex match",
+		Parameters: []interface{}{
+			getRequestNodeRegexMatch(nodeClusterField, "c.....r"),
```
Comment (jessicayuen) [inline internal/app/mapper/mapper_test.go]: Can we try some other regex variations in a few of these tests? ex: `c*`

Comment (jyotimahapatra) [reply internal/app/mapper/mapper_test.go]: sounds good. Added some more.

File: internal/app/mapper/mapper.go (hunk 2 above)
Comment (jyotimahapatra) [reply internal/app/mapper/mapper.go]: i think that is desired. Added some tests now.

**Comment by jyotimahapatra** [comment]
PTAL @jessicayuen

**Comment by jyotimahapatra** [comment]
@jessicayuen Thanks..I will work on the next ones once this gets merged.

**Pull request merged**